github/workflows: Build packages for NVIDIA platform variants(amd and arm)

yash-zededa opened the pull request:
The eve job occasionally encounters issues when building the NVIDIA variants, which may be caused by the type of runners being used.

This PR builds packages for the NVIDIA platform variants and also includes the AMD variant, as discussed with @rene , for potential future use. The eve job will then reference the packages from the cache.


Reference PR: https://github.com/lf-edge/eve/pull/4562

Execution matrix table:

OS | Architecture | Platform | Included/Excluded
-- | -- | -- | --
buildjet-4vcpu-ubuntu-2204-arm | arm64 | generic | ✅ Included
buildjet-4vcpu-ubuntu-2204-arm | arm64 | nvidia-jp5 | ✅ Included
buildjet-4vcpu-ubuntu-2204-arm | arm64 | nvidia-jp6 | ✅ Included
buildjet-4vcpu-ubuntu-2204-arm | amd64 | generic | ❌ Excluded
buildjet-4vcpu-ubuntu-2204-arm | amd64 | nvidia-jp5 | ❌ Excluded
buildjet-4vcpu-ubuntu-2204-arm | amd64 | nvidia-jp6 | ❌ Excluded
buildjet-4vcpu-ubuntu-2004 | amd64 | generic | ✅ Included
buildjet-4vcpu-ubuntu-2004 | amd64 | nvidia-jp5 | ✅ Included
buildjet-4vcpu-ubuntu-2004 | amd64 | nvidia-jp6 | ✅ Included
buildjet-4vcpu-ubuntu-2004 | arm64 | generic | ❌ Excluded
buildjet-4vcpu-ubuntu-2004 | arm64 | nvidia-jp5 | ❌ Excluded
buildjet-4vcpu-ubuntu-2004 | arm64 | nvidia-jp6 | ❌ Excluded
buildjet-4vcpu-ubuntu-2004 | riscv64 | generic | ✅ Included

rene commented on .github/workflows/build.yml:
```diff
@@ -20,13 +20,18 @@ jobs:
             arch: amd64
           - os: buildjet-4vcpu-ubuntu-2004
-            arch: riscv64
+            platform: arm64
```
There is no platform `arm64`, I think you meant here `arch: arm64` ?

yash-zededa replied:
Yup, it's a typo. I will fix it.

> There is no platform `arm64`, I think you meant here `arch: arm64` ?

yash-zededa replied:
> I assume those are exhaustive exclude patterns, right? I.e.
> 
> ```
> exclude:
>   - os: buildjet-4vcpu-ubuntu-2204-arm
>   - arch: amd64
> ```
> 
> Would assume platform: [generic, nvida-jp5, nvidia-jp6]?

Yes, those are exclude patterns that prevent cross-compilation of platforms for the OS and architecture. This means the runner `buildjet-4vcpu-ubuntu-2204-arm` won’t run on AMD, and vice versa.

rene replied:
I've missed that. In this case it's better to keep the consistence here.... let as it is or change both... TBH for me is fine to let as it is if it's being used like this across workflows sources....

rene commented on .github/workflows/build.yml:
```diff
@@ -71,9 +76,9 @@ jobs:
+      - name: Build packages ${{ matrix.os }} for ${{ matrix.platform }}
         run: |
-          make V=1 PRUNE=1 ZARCH=${{ matrix.arch }} pkgs
+          make V=1 PRUNE=1 PLATFORM=${{ matrix.platform }} ZARCH=${{ matrix.arch }} pkgs
```
if you don't mind to change, I think `make V=1 PRUNE=1 ZARCH=${{ matrix.arch }} PLATFORM=${{ matrix.platform }}` is better for readability because we always use in this order throughout the documentation...

yash-zededa replied:
@rene I don’t mind, but I just referenced the make command from the eve jobs section, and according to that, we have the following order.

```
make V=1 ROOTFS_VERSION="$VERSION" PLATFORM=${{ matrix.platform }} HV=${{ matrix.hv }} ZARCH=${{ matrix.arch }} pkgs eve
```

https://github.com/lf-edge/eve/blob/a428302f609f343310c5c700394fba088f8aaddd/.github/workflows/build.yml#L188

However, if we follow a different order, I will update it as suggested.

rene replied:
I've missed that. In this case it's better to keep the consistence here.... let as it is or change both... TBH for me is fine to let as it is if it's being used like this across workflows sources....

yash-zededa commented:
> Will that work? I always get confused by the include/exclude syntax. We currently have 3 runs:
> 
> * amd64 on buildjet-amd64
> * arm64 on buildjet-arm64
> * riscv64 on buildjet-amd64
> 
> We want:
> 
> * amd64 generic on buildjet-amd64 (same as current)
> * arm64 generic on buildjet-arm64 (same as current)
> * arm64 jetpack5 on buildjet-arm64
> * arm64 jetpack6 on buildjet-arm64
> * riscv64 generic on buildjet-amd64 (same as current)

Yes, we will have a total of five runs (three actual ones plus two NVIDIA Jetson ARM). 

Additionally, I have added two more runners—two NVIDIA Jetson and one AMD. While these won't be used immediately, they have been included for future requirements.

I have also updated the PR description to include a matrix that shows which combinations will trigger the job.

rene commented:
@yash-zededa , thanks for adding that table in the PR Description, that helps a lot!!!!

rene reviewed (approved):
LGTM

OhmSpectator reviewed (approved):
@rene, merge and test then? =)

yash-zededa commented:
@rene @OhmSpectator Let's merge and test!
